### PR TITLE
Add hd check

### PR DESCRIPTION
**Pull request by emma-p**
- Validates that the returned user's email address matches the domain we are expecting (see https://developers.google.com/identity/protocols/OpenIDConnect?hl=en#hd-param)
- Improve login controller and authentication services to make them more testable
- Adds tests

### CODE REVIEW

File: app/controllers/Login.scala [hunk 1]
```diff
@@ -39,24 +46,27 @@ class Login @Inject() (conf: Config) extends Controller with AuthActions {
...
-      case _ => {
-        val redirect = loginErrorRedirect(GroupsValidationFailed())
+      case -\/(error) => {
+        val redirect = loginErrorRedirect(error)
         redirect.withSession { session.loginError }
       }
     } recover {
       case ex => {
-        val redirect = loginErrorRedirect(IdentityValidationFailed())
+        logger.error(s"$UnexpectedError: $ex")
```
Comment (danmassie) [inline app/controllers/Login.scala]: Won't send stack trace correctly to Kibana. Use logger.error(message, throwable)

File: app/controllers/Login.scala [hunk 2]
```diff
@@ -39,24 +46,27 @@ class Login @Inject() (conf: Config) extends Controller with AuthActions {
   def oauth2Callback = CSRFAction.async { implicit request =>
```
Comment (danmassie) [inline app/controllers/Login.scala]: There should be some tests around this.

File: app/model/AdminIdentity.scala [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package model
+
+import com.gu.googleauth.UserIdentity
+
+case class AdminIdentity(userIdentity: UserIdentity)
```
Comment (danmassie) [inline app/model/AdminIdentity.scala]: Not sure what this wrapping achieves.

File: app/services/GoogleGroups.scala [hunk 4]
```diff
@@ -25,26 +29,33 @@ object GoogleGroups extends Logging {
...
-      }
-    })
-  }
+
+  def isUserAdmin[A](identity: UserIdentity): Future[\/[LoginError, AdminIdentity]] =
+    Admin.isUserInAdminGroups(userGroups, identity, requiredGroups)
 }
 
 object Authorisation {
   val TWO_FACTOR_AUTH_GROUP = "2fa_enforce@guardian.co.uk"
```
Comment (danmassie) [inline app/services/GoogleGroups.scala]: Should this be in the repo?

File: app/services/SafeGoogleAuth.scala [hunk 5]
```diff
@@ -1,21 +1,31 @@
...
-  def validatedUserIdentity[A](implicit request: CSRFRequest[A]): Future[UserIdentity] = {
-    val t = Try(GoogleAuth.validatedUserIdentity(GoogleAuthConf.googleAuthConfig, request.csrfToken)) recover {
-      case e: IllegalArgumentException => Future.failed(e)
-    }
-    t.get
+  def correctHostedDomain(identity: UserIdentity, googleAuthConfig: GoogleAuthConfig): Boolean = Some(identity.emailDomain) == googleAuthConfig.domain
+
+  def validateUserIdentity[A](f: (GoogleAuthConfig, String) => Future[UserIdentity], googleAuthConfig: GoogleAuthConfig, expectedAntiForgeryToken: String)
+                             (implicit request: CSRFRequest[A]): Future[\/[LoginError, UserIdentity]] = {
+    val t = Try(f(googleAuthConfig, expectedAntiForgeryToken).map { identity =>
```
Comment (danmassie) [inline app/services/SafeGoogleAuth.scala]: This isn't very readable.

File: test/services/GoogleGroups$Test.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import model.AdminIdentity
+import org.scalatest.{ParallelTestExecution, Matchers, FlatSpec}
+import org.scalatest.concurrent.{AsyncAssertions, ScalaFutures}
+
+import scala.concurrent.Future
+import scalaz.{-\/, \/-}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class GoogleGroups$Test extends FlatSpec with Matchers with ScalaFutures with ParallelTestExecution with AsyncAssertions {
```
Comment (danmassie) [inline test/services/GoogleGroups$Test.scala]: Fix the name to remove the $

File: test/services/GoogleGroups$Test.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  val identity = UserIdentity("sub","test.com","firstName","lastName",1234,None)
+  val adminIdentity = AdminIdentity(identity)
+  val requiredGroups = Set("validGroup")
+
+  behavior of "GoogleGroups$Test"
```
Comment (danmassie) [inline test/services/GoogleGroups$Test.scala]: $ in name again

File: app/services/GoogleGroups.scala [hunk 8]
```diff
@@ -25,26 +29,33 @@ object GoogleGroups extends Logging {
...
-    userGroups(email).map( _ match {
-      case Right(groups) => isAuthorised(required = requiredGroups, groups = groups)
-      case Left(_) => {
-        logger.info("{} is not in correct groups", email)
-        false
-      }
-    })
-  }
+
+  def isUserAdmin[A](identity: UserIdentity): Future[\/[LoginError, AdminIdentity]] =
```
Comment (danmassie) [inline app/services/GoogleGroups.scala]: The naming of this implies that a boolean will be returned.

File: app/services/GoogleGroups.scala [hunk 9]
```diff
@@ -25,26 +29,33 @@ object GoogleGroups extends Logging {
...
 
 object Authorisation {
   val TWO_FACTOR_AUTH_GROUP = "2fa_enforce@guardian.co.uk"
+
   def isAuthorised(required: Set[String], groups: Set[String]): Boolean = (required & groups) == required
 }
+
+object Admin extends Logging {
+
+  def isUserInAdminGroups(f: UserIdentity => Future[Set[String]], identity: UserIdentity, requiredGroups: Set[String]): Future[\/[LoginError, AdminIdentity]] = {
```
Comment (danmassie) [inline app/services/GoogleGroups.scala]: The naming of this implies that a boolean will be returned

File: app/controllers/Login.scala [hunk 10]
```diff
@@ -39,24 +46,27 @@ class Login @Inject() (conf: Config) extends Controller with AuthActions {
...
     val session = request.session
 
-    val result = for {
-      identity <- SafeGoogleAuth.validatedUserIdentity
-      admin <- GoogleGroups.isUserAdmin(identity.email)
-    } yield if (admin) identity
+    type FutureEither[A] = EitherT[Future, LoginError, A]
```
Comment (danmassie) [inline app/controllers/Login.scala]: Consider using https://github.com/beamly/beamly.core.scalaz for FutureEither and replacing all instances of it.

File: test/services/SafeGoogleAuth$Test.scala [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  val googleAuthConfig = GoogleAuthConfig("1", "ssh", "http://test.com", Some("test.com"), None, false)
+  val request = FakeRequest()
+  implicit val app = FakeApplication()
+  implicit val csrfAction = new CSRFRequest("token", request)
+
+  behavior of "SafeGoogleAuth$Test"
+
+  it should "validateUserIdentity" in {
+    running(app) {
+      whenReady(SafeGoogleAuth.validateUserIdentity(validUserIdentity, googleAuthConfig, "test")) { result =>
```
Comment (danmassie) [inline test/services/SafeGoogleAuth$Test.scala]: These assertions should each be in separate tests so its easy to see from the test report which one failed without looking at the stack trace

**Comment by emma-p** [comment]
:+1:

**Pull request merged**